internal/datacoord/segment_manager.go

```diff
@@ -280,38 +297,32 @@ func (s *SegmentManager) AllocSegment(ctx context.Context, collectionID UniqueID
 		With(zap.Int64("requestRows", requestRows))
 	_, sp := otel.Tracer(typeutil.DataCoordRole).Start(ctx, "Alloc-Segment")
 	defer sp.End()
-	s.mu.Lock()
-	defer s.mu.Unlock()
 
 	// filter segments
-	validSegments := make(map[UniqueID]struct{})
-	invalidSegments := make(map[UniqueID]struct{})
-	segments := make([]*SegmentInfo, 0)
-	for _, segmentID := range s.segments {
+	segmentInfos := make([]*SegmentInfo, 0)
+	partition2Segments, _ := s.channel2Segments.GetOrInsert(channelName, typeutil.NewConcurrentMap[int64, Segments]())
+	segments, _ := partition2Segments.GetOrInsert(partitionID, typeutil.NewConcurrentMap[int64, struct{}]())
+	segments.Range(func(segmentID int64, _ struct{}) bool {
 		segment := s.meta.GetHealthySegment(segmentID)
 		if segment == nil {
-			invalidSegments[segmentID] = struct{}{}
-			continue
+			log.Warn("failed to get segment, remove it", zap.String("channel", channelName), zap.Int64("segmentID", segmentID))
+			segments.Remove(segmentID)
+			return true
 		}
 
-		validSegments[segmentID] = struct{}{}
-		if !satisfy(segment, collectionID, partitionID, channelName) || !isGrowing(segment) || segment.GetLevel() == datapb.SegmentLevel_L0 {
-			continue
+		if !isGrowing(segment) || segment.GetLevel() == datapb.SegmentLevel_L0 {
+			return true
 		}
-		segments = append(segments, segment)
-	}
-
-	if len(invalidSegments) > 0 {
-		log.Warn("Failed to get segments infos from meta, clear them", zap.Int64s("segmentIDs", lo.Keys(invalidSegments)))
-	}
-	s.segments = lo.Keys(validSegments)
+		segmentInfos = append(segmentInfos, segment)
+		return true
+	})
 
 	// Apply allocation policy.
 	maxCountPerSegment, err := s.estimateMaxNumOfRows(collectionID)
 	if err != nil {
 		return nil, err
 	}
-	newSegmentAllocations, existedSegmentAllocations := s.allocPolicy(segments,
+	newSegmentAllocations, existedSegmentAllocations := s.allocPolicy(segmentInfos,
```

Comment:
do we need some lock here for each channel + partition?
otherwise under concurrent write we may allocate mulitple segments at the same timee

Reply:
Ok, use channel key lock instead